### PR TITLE
non-buffered responses fixes

### PR DESCRIPTION
**Pull request by 1ma**
- Fix potential casting from array to string in Response::withHeader()
- Fix non buffered responses when calling Response::withAddedHeader() and Response::withoutHeader()

Reference: https://github.com/slimphp/Slim-Psr7/pull/39#issuecomment-487278458

### CODE REVIEW

**Comment by l0gicgate** [comment]
Can you PR this on Slim PSR-7 as well please? Also we should add tests cases for the newly introduced code.

**Comment by 1ma** [comment]
The merged PR against slim/psr7 that I referenced in the first comment already has these fixes. Also, regarding the difficulty of testing this feature: https://github.com/slimphp/Slim/pull/2504#issuecomment-469035419

Unfortunately neither of the 2 original PRs had any (real) tests for this.

File: Slim/Http/Message.php [hunk 1]
```diff
@@ -230,6 +230,10 @@ public function withAddedHeader($name, $value)
         $clone = clone $this;
         $clone->headers->add($name, $value);
 
+        if ($this instanceof Response && $this->body instanceof NonBufferedBody) {
```
Comment (l0gicgate) [inline Slim/Http/Message.php:233]: We can add tests by overriding the `header()` can't we?

File: Slim/Http/Message.php [hunk 2]
```diff
@@ -251,6 +255,10 @@ public function withoutHeader($name)
         $clone = clone $this;
         $clone->headers->remove($name);
 
+        if ($this instanceof Response && $this->body instanceof NonBufferedBody) {
+            header_remove($name);
```
Comment (l0gicgate) [inline Slim/Http/Message.php:259]: We can also override `header_remove()` here just to make sure it's being called as well.

File: Slim/Http/Response.php [hunk 3]
```diff
@@ -270,13 +270,13 @@ public function getReasonPhrase()
...
     {
-        if ($this->body instanceof NonBufferedBody) {
-            header("$name: $value");
-        }
-
         $clone = clone $this;
         $clone->headers->set($name, $value);
 
+        if ($this->body instanceof NonBufferedBody) {
+            header(sprintf('%s: %s', $name, $this->getHeaderLine($name)));
```
Comment (l0gicgate) [inline Slim/Http/Response.php]: Same here we can override `header()`.

File: Slim/Http/Message.php (hunk 1 above)
Comment (1ma) [reply Slim/Http/Message.php:233]: I don't know how to do that and Google isn't shedding much light either. Can you think of any example?

Comment (1ma) [reply Slim/Http/Message.php:233]: Unless you mean moving the header and header_remove functions to new private methods just for the sake of mocking them (in which case code coverage would still be incomplete)?

**Comment by 1ma** [comment]
That'd be it. I've done some cleanup in the testing autoloading, too. These overrides should only be loaded from tests/bootstrap.php, not at each `composer install`. Now to port these new tests back to Slim-Psr7...

**Review by l0gicgate — APPROVED** [review]

**Comment by l0gicgate** [comment]
I'm somewhat worried we might break some builds in the wild as seen in #2669 because we're moving the assets into `bootstrap.php` but I guess we'll have to deal with it when the issues come in. Great work @1ma!

**Pull request merged**